File: C++/Array.cpp
/*
      Author : Prabhat kumar
      date : 15-jan-2020
      
Design, Develop and Implement a menu driven Program for the following Array operations
a. Creating an Array of N Integer Elements
b. Display of Array Elements with Suitable Headings
c. Inserting an Element (ELEM) at a given valid Position (POS)
d. Deleting an Element at a given valid Position(POS)
e. Exit

In addition it is resizeable but not like vector it size gets twice. In this the array size get new user define size

*/

#include <iostream>
#include <cstdlib>
#include <string.h>
#include <stdio.h>
using namespace std;

class Array{
  public:
  int *arr;
  int size;
  Array(int si=10){
    arr=new int[si];
    size=si;
  }
  void print(){
    for(int i=0; i<size; i++)
    cout<<"index : "<<i<<" | data : "<<arr[i]<<endl;
  }
  void insert(){ 
    int pos=0, data=0;
    char confermation='y';
    while(confermation=='y'){
      cout<<"Enter data to insert : ";
      cin>>data;
      cout<<"Enter position to insert : ";
      cin>>pos;
      if(pos>size)cout<<"Invalid position"<<endl;
      else arr[pos]=data;
      cout<<"do you want to insert more[y/n] : ";
      cin>>confermation;
      confermation=tolower(confermation);
    }
  }
  void Delete(){
    int pos=0;
    char confermation='y';
    while(confermation=='y'){
      cout<<"Enter position to Delete : ";
      cin>>pos;
      int i=0;
      for(i=pos; i<size-1; i++)arr[i]=arr[i+1];
      arr[i]=0;
      cout<<"do you want to Delete more[y/n] : ";
      cin>>confermation;
      confermation=tolower(confermation);
    }
  }
  void resize(){
    cout<<"Enter the size of an Array "<<endl;
    size_t new_size;
    cin>>new_size;
     int* newArr = new int[new_size];
    memcpy( newArr, arr, new_size * sizeof(int) );
   delete[] arr;
   arr = newArr;
  for(int i=size; i<size+3; i++)arr[i]=0;
   size=new_size;
  }
  void Size(){
    cout<<"size of Array is : "<<size<<endl;
  }
};
void switch_case(int n,Array* b){
   switch(n){
     case 1:
     b->print();
     break;
     case 2:
     b->insert();
     break;
     case 3:
     b->Delete();
     break;
     case 4:
     b->Size();
     break;
     case 5:
     b->resize();
     break;
     case 6:
     exit(0);
     default: cout<<"Enter a valid number "<<endl;
   }
}

int main() {
  int size;
  cout<<"Enter size of array : ";
  cin>>size;
  Array arr(size);
  while(1){
    int options;
    cout<<"\t1. Print the Array "<<endl;
    cout<<"\t2. Insert data in Array "<<endl;
    cout<<"\t3. Delete data in Array "<<endl;
    cout<<"\t4. Show the size of the array"<<endl;
    cout<<"\t5. Resize the Array "<<endl;
    cout<<"\t6. Exit "<<endl;
    cout<<"Enter the option : ";cin>>options;
    switch_case(options, &arr);
    getchar();
    getchar();
    system("clear");
  }
}

File: Python/Array.py
'''
            author : Prabhat kumar
            date : 28 jan 2020
        Fully dynamic array implemented in python
        Design, Develop and Implement a menu driven Program for the following Array operations
a. Creating an Array of N Integer Elements
b. Display of Array Elements with Suitable Headings
c. Inserting an Element (ELEM) at a given valid Position (POS)
d. Deleting an Element at a given valid Position(POS)
e. Exit
'''
from os import system
class Array:
    def __init__(self,size):
        self.size=size;
        self.arr=[0 for i in range(size)]
    def print_all(self):
        for i in range(len(self.arr)):
            print("Index ",i+1,"| data : ",self.arr[i])
    def insert(self):
        data=int(input("Enter the data to insert : "))
        pos=int(input("Enter position to insert : "))
        if self.arr[pos-1]!=0:
            print("Already filled! do you want to replace[y/n] : ",end="")
            confermation=input()
            if confermation=="y":
                self.arr[pos-1]=data
                return
        self.arr[pos-1]=data
    def delete(self):
        pos=int(input("Enter position to delete"))
        if pos-1>=size or self.arr[pos-1]==0:
            print("Cannot delete position out of bound or there is no data")
        else:
            self.arr[pos-1]=0
    def resize(self):
        num=int(input("Enter new size of arr "))
        if num<len(self.arr):
            self.arr=self.arr[:num]
        else:
            self.arr=self.arr+[0 for i in range(num-len(self.arr))]
        
if __name__=="__main__":
    size=int(input("Enter the size of array "))
    array=Array(size)
    while True:
        print("\t1. Print the Array ")
        print("\t2. Insert data in Array ")
        print("\t3. Delete data in Array ")
        print("\t4. Show the size of the array")
        print("\t5. Resize the Array ")
        print("\t6. Exit ")
        options=int(input("Choose from above(1-6) : "))
        if options==1:
            array.print_all()
        elif options==2:
            array.insert()
        elif options==3:
            array.delete()
        elif options==5:
            array.resize()
        elif options==4:
            print("Size of arra is ",array.size)
        elif options==6:
            break
        else:
            print("Wrong options selected \n Choose right option")
        input("Press Enter to continue")
        system("clear")
    
